### ISSUE TITLE
gatsby-source-drupal - allow json api filter parameters

### ISSUE THREAD

**Issue by apmsooner** [gatsbyjs/gatsby#11696]
jsonapi module allows query parameters to filter list like: /jsonapi?filter[some_field]=some_value

It would be nice option if gatsby-source-drupal could accommodate this so one drupal site could store data for a number of gatsby sites and the gatsby sites could just get the data they need. An example would be a filtering nodes by a taxonomy term so it might mean gatsby processing 100 nodes instead of 10,000.

**Comment by pieh** [comment]
How would that work - should this apply same query params to all requests?

Right now, we hit main endpoint to get list of available endpoints and then make requests for those - should all of those requests have same query params? Would this potentially cause using taxonomy term filter for endpoint where it doesn't make sense (i.e. on content type that doesn't use this taxonomy) and potentially "nulling" the output of it? (I don't have enough context on drupal jsonapi module)

Just for testing the idea you could "monkey patch" `node_modules/gatsby-source-drupal/gatsby-node.js` file and find `axios.get(<some_path>, { auth: basicAuth })` (should be 2 of them) and add `params` there:

`axios.get(<some_path>, { auth: basicAuth, params: { "filter[some_field]": "some_value } })`

and see what happens

**Comment by apmsooner** [comment]
Yeah, I was thinking in gatsby-config.js in line with how you wrote out the example being able to specify:
```
resolve: `gatsby-source-drupal`,
      options: {
        baseUrl: `http://somesite.com/`,
        apiBase: `jsonapi`,
        parameters: {},
        basicAuth: {},
      },
```
If the filter field isn't present or the value isn't found, yeah... null the results. I'll look into trying what you described. Seems like it should be doable and would really speed up the build process when dealing with a backend site containing alot of data.

**Comment by apmsooner** [comment]
Oh... wait now i know what you are saying cause it would affect ALL endpoints within the jsonapi path. I saw another issue request to ask for multiple json api endpoints. I think that would have to happen first make this doable.

**Comment by pieh** [comment]
Just to check we are on the same page I'll walk you through how currently drupal plugin works:

 1. First we hit `http://domain.example/jsonapi` which give us list of dedicated endpoints for each content types / taxonomies and so on - response looks something like this: https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-source-drupal/src/__tests__/fixtures/jsonapi.json
 2. Then we iterate over list of endpoints from this reponse and do request for each - using above example we would hit endpoints for `file--file`, `node--article` and `taxonomy_term--tags` (and from those we actually get data to create Gatsby nodes).

So the question is if using `{ "filter[some_field]": "some_value" }` query param on each of those requests make sense? It might for `node--article`, but it might not for `file--file`, right?

what about this design:
```js
{
	resolve: `gatsby-source-drupal`,
    options: {
        baseUrl: `http://somesite.com/`,
        parameters: type => {
          if (type === `node--article`) {
            return { "filter[some_field]": "some_value" }
          }
          // for everything else - no params
          return {}
        },
    },
}
```

I'm not sure how linked issue would exactly help in this case, but it would be great to fix it separately too for sure.

**Comment by apmsooner** [comment]
Yeah we're on the same page and agree with you on not restricting request by param as it would cause issues between different entity types. Your design proposal seems like a really good solution.

**Comment by apmsooner** [comment]
@pieh, 
I went a slightly different route but I've been able to make it work without any errors. Not sure its the best way as I'm primarily a php coder but I'd like to get your feedback.

Structure for gatsby-config.js

`resolve: `gatsby-source-drupal`,
      options: {
        baseUrl: `http://mysite.com/`,
        apiBase: `jsonapi`,
        filters: {
          filter: {
            type: `foods`,
            parameter: `filter[ndb][condition][path]=name&filter[ndb][condition][operator]=STARTS_WITH&filter[ndb][condition][value]=apple`
          }
        }
      },`

Restructured gatsby-node.js (from gatsby-source-drupal)
`"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

const axios = require(`axios`);

const crypto = require(`crypto`);

const _ = require(`lodash`);

const _require = require(`gatsby-source-filesystem`),
      createRemoteFileNode = _require.createRemoteFileNode;

const _require2 = require(`url`),
      URL = _require2.URL;

const _require3 = require(`./normalize`),
      nodeFromData = _require3.nodeFromData; // Get content digest of node.


const createContentDigest = obj => crypto.createHash(`md5`).update(JSON.stringify(obj)).digest(`hex`);

exports.sourceNodes =
/*#__PURE__*/
function () {
  var _ref = (0, _asyncToGenerator2.default)(function* ({
    actions,
    getNode,
    hasNodeChanged,
    store,
    cache,
    createNodeId
  }, {
    baseUrl,
    apiBase,
    filters,
    basicAuth
  }) {
    const createNode = actions.createNode; // Default apiBase to `jsonapi`

    apiBase = apiBase || `jsonapi`; // Touch existing Drupal nodes so Gatsby doesn't garbage collect them.
    // _.values(store.getState().nodes)
    // .filter(n => n.internal.type.slice(0, 8) === `drupal__`)
    // .forEach(n => touchNode({ nodeId: n.id }))
    // Fetch articles.
    // console.time(`fetch Drupal data`)

    console.log(`Starting to fetch data from Drupal`); // TODO restore this
    // let lastFetched
    // if (
    // store.getState().status.plugins &&
    // store.getState().status.plugins[`gatsby-source-drupal`]
    // ) {
    // lastFetched = store.getState().status.plugins[`gatsby-source-drupal`].status
    // .lastFetched
    // }

    const data = yield axios.get(`${baseUrl}/${apiBase}`, {
      auth: basicAuth
    });
    const allData = yield Promise.all(_.map(data.data.links,
    /*#__PURE__*/
    function () {
      var _ref2 = (0, _asyncToGenerator2.default)(function* (url, type) {
        if (type === `self`) return;
        if (!url) return;
        if (!type) return;

        const getNext =
        /*#__PURE__*/
        function () {
          var _ref3 = (0, _asyncToGenerator2.default)(function* (url, data = []) {
            if (typeof url === `object`) {
              // url can be string or object containing href field
              url = url.href;
              if (filters.filter.type === type) {
                url = url + `?${filters.filter.parameter}`;
              }
            }

            let d;

            try {
              d = yield axios.get(url, {
                auth: basicAuth
              });
            } catch (error) {
              if (error.response && error.response.status == 405) {
                // The endpoint doesn't support the GET method, so just skip it.
                return [];
              } else {
                console.error(`Failed to fetch ${url}`, error.message);
                console.log(error.data);
                throw error;
              }
            }

            data = data.concat(d.data.data);

            if (d.data.links.next) {
              data = yield getNext(d.data.links.next, data);
            }

            return data;
          });

          return function getNext(_x5) {
            return _ref3.apply(this, arguments);
          };
        }();

        const data = yield getNext(url);
        const result = {
          type,
          data // eslint-disable-next-line consistent-return

        };
        return result;
      });

      return function (_x3, _x4) {
        return _ref2.apply(this, arguments);
      };
    }())); // Make list of all IDs so we can check against that when creating
    // relationships.

    const ids = {};

    _.each(allData, contentType => {
      if (!contentType) return;

      _.each(contentType.data, datum => {
        ids[datum.id] = true;
      });
    }); // Create back references


    const backRefs = {};
    /**
     * Adds back reference to linked entity, so we can later
     * add node link.
     */

    const addBackRef = (linkedId, sourceDatum) => {
      if (ids[linkedId]) {
        if (!backRefs[linkedId]) {
          backRefs[linkedId] = [];
        }

        backRefs[linkedId].push({
          id: sourceDatum.id,
          type: sourceDatum.type
        });
      }
    };

    _.each(allData, contentType => {
      if (!contentType) return;

      _.each(contentType.data, datum => {
        if (datum.relationships) {
          _.each(datum.relationships, (v, k) => {
            if (!v.data) return;

            if (_.isArray(v.data)) {
              v.data.forEach(data => addBackRef(data.id, datum));
            } else {
              addBackRef(v.data.id, datum);
            }
          });
        }
      });
    }); // Process nodes


    const nodes = [];

    _.each(allData, contentType => {
      if (!contentType) return;

      _.each(contentType.data, datum => {
        const node = nodeFromData(datum, createNodeId);
        node.relationships = {}; // Add relationships

        if (datum.relationships) {
          _.each(datum.relationships, (v, k) => {
            if (!v.data) return;

            if (_.isArray(v.data) && v.data.length > 0) {
              // Create array of all ids that are in our index
              node.relationships[`${k}___NODE`] = _.compact(v.data.map(data => ids[data.id] ? createNodeId(data.id) : null));
            } else if (ids[v.data.id]) {
              node.relationships[`${k}___NODE`] = createNodeId(v.data.id);
            }
          });
        } // Add back reference relationships.
        // Back reference relationships will need to be arrays,
        // as we can't control how if node is referenced only once.


        if (backRefs[datum.id]) {
          backRefs[datum.id].forEach(ref => {
            if (!node.relationships[`${ref.type}___NODE`]) {
              node.relationships[`${ref.type}___NODE`] = [];
            }

            node.relationships[`${ref.type}___NODE`].push(createNodeId(ref.id));
          });
        }

        if (_.isEmpty(node.relationships)) {
          delete node.relationships;
        }

        node.internal.contentDigest = createContentDigest(node);
        nodes.push(node);
      });
    }); // Download all files.


    yield Promise.all(nodes.map(
    /*#__PURE__*/
    function () {
      var _ref4 = (0, _asyncToGenerator2.default)(function* (node) {
        let fileNode;

        if (node.internal.type === `files` || node.internal.type === `file__file`) {
          try {
            let fileUrl = node.url;

            if (typeof node.uri === `object`) {
              // Support JSON API 2.x file URI format https://www.drupal.org/node/2982209
              fileUrl = node.uri.url;
            } // Resolve w/ baseUrl if node.uri isn't absolute.


            const url = new URL(fileUrl, baseUrl); // If we have basicAuth credentials, add them to the request.

            const auth = typeof basicAuth === `object` ? {
              htaccess_user: basicAuth.username,
              htaccess_pass: basicAuth.password
            } : {};
            fileNode = yield createRemoteFileNode({
              url: url.href,
              store,
              cache,
              createNode,
              createNodeId,
              auth
            });
          } catch (e) {// Ignore
          }

          if (fileNode) {
            node.localFile___NODE = fileNode.id;
          }
        }
      });

      return function (_x6) {
        return _ref4.apply(this, arguments);
      };
    }()));
    nodes.forEach(n => createNode(n));
  });

  return function (_x, _x2) {
    return _ref.apply(this, arguments);
  };
}();`

**Comment by apmsooner** [comment]
Sorry, my paste into markdown is not too legible. The main change to gatsby-node.js is:

```
if (filters.filter.type === type) {
            url = url + `?${filters.filter.parameter}`;
          }
```

Again i don't know if this is the optimal way of doing it but it works like a charm and filters exactly like i needed to. And this way, i can combine various other filters on other node types.

**Comment by eojthebrave** [comment]
Adding  the ability to provide per resource type filters seems like a good approach. And is something I would be able to make use of it it existed.

@apmsooner any chance you could create a pull request with this feature added? I would be happy to help review and test code if you could.

**Comment by apmsooner** [comment]
@eojthebrave, 
Here's attached gatsby-node.js file. It was only a couple lines of code changed. Just drop that in /node_modules/gatsby-source-drupal. Then in gatsby-config.js... something like this example where 'type' is the type of node to affect and parameter is the query parameter string to filter on excluding the ? of course as its accounted for. Again, I don't know if its the right approach but it works for me... I'd appreciate feedback.

`resolve: `gatsby-source-drupal`,
      options: {
        baseUrl: `http://yoursite.com/`,
        apiBase: `jsonapi`
        basicAuth: {
          username: process.env.BASIC_AUTH_USERNAME,
          password: process.env.BASIC_AUTH_PASSWORD,
        },
        filters: {
          filter: {
            type: `hotels`,
            parameter: `include=brand_id&filter[status]=1&filter[brand][condition][path]=brand_id.brand_code&filter[brand][condition][operator]=IN&filter[brand][condition][value][]=chi&filter[brand][condition][value][]=abi&filter[brand][condition][value][]=3ph&filter[brand][condition][value][]=jam`
          }
        }
      },`
[gatsby-node.js.zip](https://github.com/gatsbyjs/gatsby/files/2934219/gatsby-node.js.zip)

**Comment by apmsooner** [comment]
Sorry, i can't get markdown to retain my pasted code as formatted. I must be doing something wrong...

**Comment by apmsooner** [comment]
Alright, i figured out how to fix the formatting of the code i pasted so this should all make a whole lot more sense now :)

**Comment by apmsooner** [comment]
Here's another example making use of drupal node types with multiple filters:

```
{
      resolve: `gatsby-source-drupal`,
      options: {
        baseUrl: `http://yoursite.com/`,
        apiBase: `jsonapi`,
        basicAuth: {
          username: process.env.BASIC_AUTH_USERNAME,
          password: process.env.BASIC_AUTH_PASSWORD,
        },
        filters: {
          filter: {
            type: `node--page`,
            parameter: `filter[status]=1`
          },
          filter: {
            type: `node--article`,
            parameter: `filter[condition][path]=title&filter[condition][operator]=STARTS_WITH&filter[condition][value]=the`
          }
        }
      },
    }
```

**Comment by apmsooner** [comment]
After playing around with multiple filters, i think the logic is wrong and need to restructure to ensure operating on the right key. This is what I think might be best.

gatsy-source-drupal/gatsby-node.js
```
if (typeof url === `object`) {
              // url can be string or object containing href field
              url = url.href;
              if (typeof filters === `object`) {
                if (filters.hasOwnProperty(type)) {
                  url = url + `?${filters[type]}`;
                }
              }
            }
```
gatsby-config.js example:

```
resolve: `gatsby-source-drupal`,
      options: {
        baseUrl: `http://rlhcoredev.prod.acquia-sites.com/`,
        apiBase: `jsonapi`,
        basicAuth: {
          username: process.env.BASIC_AUTH_USERNAME,
          password: process.env.BASIC_AUTH_PASSWORD,
        },
        filters: {
          "hotels": `include=brand_id&filter[status]=1&filter[brand][condition][path]=brand_id.brand_code&filter[brand][condition][operator]=IN&filter[brand][condition][value][]=chi&filter[brand][condition][value][]=abi&filter[brand][condition][value][]=3ph&filter[brand][condition][value][]=jam`,
          "brands": `filter[condition][path]=name&filter[condition][operator]=CONTAINS&filter[condition][value]=best`
        }
      },
```

This change worked perfectly for me on both filters. Suggestions?

**Comment by eojthebrave** [comment]
This is working great for me. Thanks @apmsooner. I created a pull request with the code from your example above and some updated documentation here https://github.com/gatsbyjs/gatsby/pull/12508 if you want to give it a review.

**Comment by apmsooner** [comment]
I gave it a review. Thanks Joe! I'm happy it tested out good for you and think it will be helpful to others as well particularly with large data sets.

**Comment by sidharthachatterjee** [comment]
Thank you for the great work @apmsooner and @eojthebrave 👍 

This just got published in gatsby-source-drupal@3.0.28
